File: src/com/github/doomkin/jsubsetsum/SubsetSumSolver.java
/**
 * Class SubsetSumSolver implements parallel statistical solver of 
 * subset sum problem. The existence of solution is checked by 
 * Dynamic Programming.
 * 
 * @author Pavel Nikitin
 * License: MIT
 */
package com.github.doomkin.jsubsetsum;

import java.io.PrintStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.concurrent.Callable;

public class SubsetSumSolver implements Callable<ArrayList<Integer>>
{
    private List<Integer> _list;
    private int _sum;
    private PrintStream _log;
    
    public SubsetSumSolver(List<Integer> list, int sum, PrintStream log)
    {
        _list = new ArrayList<Integer>(list);
        _sum = sum;
        _log = log;
    }
    
    public ArrayList<Integer> call() 
    {
        Date start = new Date();

        // calculate the sum of elements of list
        int totalSum = 0;
        for (int integer : _list)
            totalSum += integer;
        
        // calculate the average
        int average = totalSum / _list.size();
        
        // calculate the average subset size
        int averageSize = _sum / average;
        
        // check solution existence by Dynamic Programming
        if (averageSize < 1 || averageSize > _list.size() ||
                CombinationsSequence.lg2(_sum) > 31 || 
                !isSubsetSumDP(_list, _sum))
        {
            if (_log != null)
                _log.println("\nNo solution in " + Program.duration(start, new Date()));

            return null;
        }
            
        if (_log != null)
            _log.println("");

        // sort the list in ascending order of deviations from the average
        Collections.sort(_list, new AverageComparator(average));

        // creates computational processes for enumerating k-combinations
        TaskQueue<Combination> tasks = null;
        ArrayList<Integer> solution = null;
        
        // solve the problem
        try 
        {
            tasks = new TaskQueue<Combination>(
                    Runtime.getRuntime().availableProcessors());
            
            // processes are created in ascending order of deviations from 
            // the average size of the desired subset
            for (int size : new SubsetSizesSequence(_list.size(), averageSize))
                tasks.addLast(new FixedSizeSolver(_list, _sum, size, _log));
            
            // find combination for the solution
            Combination combination = tasks.invokeAny();
            
            // collect the solution
            if (combination != null)
            {
                solution = new ArrayList<Integer>(combination.size());
                for (int index : combination)
                    solution.add(_list.get(index));
            }
        }
        finally 
        {
            tasks.close();
        }
        
        if (_log != null)
        {
            if (solution != null)
                _log.println("\n" + solution.size() + " items found in " + 
                        Program.duration(start, new Date()));
            else
                _log.print("No solution in " + 
                        Program.duration(start, new Date()));
        }
        
        return solution;
    }
    
    // Dynamic Programming of Subset Sum Problem
    private boolean isSubsetSumDP(List<Integer> list, int sum)
    {
        if (_log != null)
            _log.print("Check solution existence");
        
        // the value of subset[i][j] will be true 
        // if there is a subset of set[0..j-1] with sum equal to i
        int n = list.size();
        BitMatrix subset = new BitMatrix(sum + 1, n + 1);
     
        // if sum is 0, then answer is true
        for (int i = 0; i <= n; i++)
            subset.set(0, i, true);
     
        // if sum is not 0 and set is empty, then answer is false
        for (int i = 1; i <= sum; i++)
            subset.set(i, 0, false);
     
        // fill the subset table
        for (int i = 1, step = Math.max(sum / 10, 1); i <= sum; i++)
        {
            if (_log != null && i % step == 0)
                _log.print(".");
            
            for (int j = 1; j <= n; j++)
            {
                subset.set(i, j, subset.get(i, j - 1));
                if (i >= list.get(j - 1))
                    subset.set(i, j, 
                            subset.get(i, j) || 
                            subset.get(i - list.get(j - 1), j - 1));
             }
        }
        
        if (_log != null)
            _log.println("");
     
        return subset.get(sum, n);
    }
}

File: src/SubsetSumSolver.kt
/** Параллельный статистический алгоритм для решения задачи о сумме подмножества */

import java.io.PrintStream
import java.util.ArrayList
import java.util.Collections
import java.util.concurrent.Callable

import java.lang.Integer.max

class SubsetSumSolver(list: List<Int>, private val sum: Int, private val log: PrintStream?) :
  Callable<ArrayList<Int>> {
  private val list: List<Int> = ArrayList(list)

  override fun call(): ArrayList<Int>? {
    var timer = Timer(log)

    // Сумма элементов исходного набора
    var totalSum = 0
    for (integer in list)
        totalSum += integer

    // Среднее арифметическое исходного набора
    val average = totalSum / list.size

    // Средний размер искомого подмножества
    val averageSize = max(sum / average, 1)

    // Проверка существования решения
    if (averageSize < 1 || averageSize > list.size ||
      CombinationsSequence.log2(sum.toLong()) > 31 ||
      !solutionExists(list, sum)
    ) {
      log?.println(" Решения не существует ($timer)")
      return null
    }

    log?.println(" OK ($timer)\n")
    timer = Timer(log)

    // Сортировка исходного набора в порядке возрастания отклонения от среднего
    Collections.sort(list, AverageComparator(average))

    // Создание процессов для перечисления k-комбинаций
    var tasks: TaskQueue<Combination>? = null
    var solution: ArrayList<Int>? = null

    // Решение задачи
    try {
      tasks = TaskQueue(Runtime.getRuntime().availableProcessors())

      // Процессы созданы в порядке возрастания отклонения от среднего размера искомого подмножества
      for (size in SubsetSizesSequence(list.size, averageSize))
        tasks.addLast(FixedSizeSolver(list, sum, size, log))

      // Поиск решающей задачу комбинации
      val combination = tasks.invokeAny()

      // Формирование решения
      if (combination != null) {
        solution = ArrayList(combination.size)
        for (index in combination)
          solution.add(list[index])
      }
    } finally {
      tasks!!.close()
    }

    if (log != null) {
      if (solution != null)
        log.println("\nНайдено ${solution.size} элементов ($timer)")
      else
        log.print("Решения не существует ($timer)")
    }

    return solution
  }

  // Проверка существования решения методом динамического программирования
  private fun solutionExists(list: List<Int>, sum: Int): Boolean {
    log?.print("Проверка существования решения")

    // Значение subset[i][j] истинно, если существует подмножество subset[0..j-1] с суммой, равной i
    val n = list.size
    val subset = BitMatrix(sum + 1, n + 1)

    // Для нулевой суммы есть решение - пустое подмножество
    for (i in 0..n)
      subset[0, i] = true

    // Для пустого исходного набора решения нет
    for (i in 1..sum)
      subset[i, 0] = false

    // Заполнение таблицы подмножеств
    var i = 1
    val step = max(sum / 10, 1)
    while (i <= sum) {
      if (log != null && i % step == 0)
        log.print(".")

      for (j in 1..n) {
        subset[i, j] = subset[i, j - 1]
        if (i >= list[j - 1])
          subset[i, j] = subset[i, j] || subset[i - list[j - 1], j - 1]
      }
      i++
    }

    return subset[sum, n]
  }
}